### PR TITLE
feat: manage default uri plugins for DevWorkspaces

### PR DESCRIPTION
**Pull request by dkwon17**
Signed-off-by: David Kwon <dakwon@redhat.com>

### What does this PR do?
This PR is the same PR as https://github.com/eclipse-che/che-dashboard/pull/425, except, its source branch is in the `eclipse-che/che-dashboard` repository rather than `dkwon17/che-dashboard`.

### What issues does this PR fix or reference?
eclipse/che#20090

### CODE REVIEW

**Review by ibuziuk — APPROVED** [review]

**Review by amisevsk — APPROVED** [review]

**Comment by dmytro-ndp** [comment]
[dashboard-ci-test]

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts [hunk 1]
```diff
@@ -273,6 +273,10 @@ export const actionCreators: ActionCreators = {
         } else {
           updatedWorkspace = await devWorkspaceClient.changeWorkspaceStatus(workspace, true, true);
         }
+        await devWorkspaceClient.onStart(
+          updatedWorkspace,
+          getState().dwPlugins.defaultEditorName as string,
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts:278]: I'm not sure if the default editor will always fit here. If I got it correctly, you need somehow to figure out which che-editor was used on the workspace creation. Or, what may be easier, store che-editor as devworkspace annotation when it's created, so then you'll be able to get it on the devworkspace start.

File: packages/dashboard-backend/src/devworkspace-client/services/api/serverConfigApi.ts [hunk 2]
```diff
@@ -21,8 +21,8 @@ const GROUP = 'org.eclipse.che';
 const VERSION = 'v1';
 const PLURAL = 'checlusters';
 
-const NAME = process.env.CHE_CRD_OBJECT_NAME;
-const NAMESPACE = process.env.CHE_NAMESPACE;
+const NAME = process.env.CHECLUSTER_CR_NAME;
+const NAMESPACE = process.env.CHECLUSTER_CR_NAMESPACE;
```
Comment (dkwon17) [inline packages/dashboard-backend/src/devworkspace-client/services/api/serverConfigApi.ts:25]: Renamed to match env variable names from the che-operator:  https://github.com/dkwon17/che-operator/blob/051d7bd9d30d1b48a9752b095426c39f2b19aee1/pkg/deploy/dashboard/deployment_dashboard.go#L59-L64

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts (hunk 1 above)
Comment (dkwon17) [reply packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts]: > If I got it correctly, you need somehow to figure out which che-editor was used on the workspace creation

Yes, that was the idea, and sounds good, my latest push sets an annotation for the che-editor ID:
https://github.com/eclipse-che/che-dashboard/blob/ebf10c803c3a80d1ecf511a146718bcdee0981c4/packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts#L466-L471

image

File: packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.onStart.spec.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+import { container } from '../../../../inversify.config';
+import { DevWorkspaceBuilder } from '../../../../store/__mocks__/devWorkspaceBuilder';
+import { DevWorkspaceClient } from '../devWorkspaceClient';
+import * as DwApi from '../../../dashboard-backend-client/devWorkspaceApi';
+import * as ServerConfigApi from '../../../dashboard-backend-client/serverConfigApi';
+
```
Comment (akurinnoy) [inline packages/dashboard-frontend/src/services/workspace-client/devworkspace/__tests__/devWorkspaceClient.onStart.spec.ts:18]:
```suggestion

/* eslint-disable @typescript-eslint/no-non-null-assertion 
```

**Review by akurinnoy — APPROVED** [review]

File: packages/dashboard-frontend/src/services/dashboard-backend-client/serverConfigApi.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * Returns an array of default plug-ins for the specified editor
+ *
+ * @param editorId The editor id to get default plug-ins for
+ * @returns Promise resolving with the array of default plug-ins for the specified editor
+ */
+export async function getDefaultPlugins(editorId: string): Promise<string[]> {
+  const url = `${prefix}/server-config/default-plugins`;
+  try {
+    const response = await axios.get(url);
+    const editorPlugins = response.data.find(
```
[the remark this answers is not in the document]
Comment (dkwon17) [reply packages/dashboard-frontend/src/services/dashboard-backend-client/serverConfigApi.ts:27]: I see. For this approach, if the list of default plugins changes in the Che CR, is it possible to receive the changes on the dashboard side without restarting the dashboard?

[the remark this answers is not in the document]
Comment (olexii4) [reply packages/dashboard-frontend/src/services/dashboard-backend-client/serverConfigApi.ts]: It is about receiving the changes on the refreshing dashboard page.
If you want to have default plugins changed immediately, you need to create a watcher, as we have for namespaces 
https://github.com/eclipse-che/che-dashboard/blob/main/packages/dashboard-frontend/src/services/bootstrap/PreloadData.ts#L115

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts [hunk 5]
```diff
@@ -143,11 +143,15 @@ export type ActionCreators = {
...
-
+// updateDevWorkspaceStatus,
+// updateDeletedDevWorkspaces,
+// updateAddedDevWorkspaces,
+// requestWorkspaces,
 export const actionCreators: ActionCreators = {
   updateAddedDevWorkspaces:
     (workspaces: devfileApi.DevWorkspace[]): AppThunk<KnownAction, void> =>
     (dispatch): void => {
+      console.log('updateAddedDevWorkspaces');
```
Comment (olexii4) [inline packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts:154]: Do we really need this log 'updateAddedDevWorkspaces'?

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts [hunk 6]
```diff
@@ -159,6 +163,7 @@ export const actionCreators: ActionCreators = {
   updateDeletedDevWorkspaces:
     (deletedWorkspacesIds: string[]): AppThunk<KnownAction, void> =>
     (dispatch): void => {
+      console.log('updateDeletedDevWorkspaces');
```
Comment (olexii4) [inline packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts:166]: Do we really need this log 'updateDeletedDevWorkspaces'?

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts [hunk 7]
```diff
@@ -170,12 +175,14 @@ export const actionCreators: ActionCreators = {
   updateDevWorkspaceStatus:
     (message: IStatusUpdate): AppThunk<KnownAction, void> =>
     (dispatch): void => {
+      console.log('updateDevWorkspaceStatus');
```
Comment (olexii4) [inline packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts:178]: Do we really need this log 'updateDevWorkspaceStatus'?

File: packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts [hunk 8]
```diff
@@ -170,12 +175,14 @@ export const actionCreators: ActionCreators = {
...
       onStatusUpdateReceived(dispatch, message);
     },
 
   requestWorkspaces:
     (): AppThunk<KnownAction, Promise<void>> =>
     async (dispatch, getState): Promise<void> => {
+      console.log('requestWorkspaces');
```
Comment (olexii4) [inline packages/dashboard-frontend/src/store/Workspaces/devWorkspaces/index.ts:185]: Do we really need this log 'requestWorkspaces'?

**Comment by dkwon17** [comment]
Thank you for the reviews everyone, I made a new commit that sets the default plugins in the redux store, and adds tests

File: packages/dashboard-frontend/src/services/dashboard-backend-client/serverConfigApi.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dkwon17) [reply packages/dashboard-frontend/src/services/dashboard-backend-client/serverConfigApi.ts]: Sounds good, I have created an issue for here: https://github.com/eclipse/che/issues/20960

**Review by olexii4 — APPROVED** [review]
LGTM

**Comment by dmytro-ndp** [comment]
[dashboard-ci-test]

**Comment by musienko-maxim** [comment]
I rechecked the image from this PR locally the problem with the Happy Path test did not reproduce. We can merge this PR

**Pull request merged**